cpp/oneapi/dal/test/engine/fixtures.hpp

```diff
@@ -230,11 +241,65 @@ class crtp_algo_fixture : public float_algo_fixture<std::tuple_element_t<0, Test
 
         return this->merge_compute_result(results);
     }
+};
 
-private:
-    Derived& derived() {
-        return *(static_cast<Derived*>(this));
+#ifdef ONEDAL_DATA_PARALLEL
+
+struct ccl_backend {};
+struct mpi_backend {};
+
+template <typename TestType, typename Derived>
+class crtp_par_algo_fixture : public crtp_base_algo_fixture<TestType, Derived> {
+    using crtp_base_algo_fixture<TestType, Derived>::derived;
+
+public:
+    template <typename... Args>
+    auto get_comm(Args&&... args) {
+        derived().get_comm(std::forward<Args>(args)...);
+    }
+    template <typename Descriptor, typename... Args>
+    auto train_in_parallel(const Descriptor& desc, Args&&... args) {
+        auto comm = derived().get_comm();
+        auto policy = this->get_policy();
+        const auto input_per_rank =
+            this->split_train_input(comm.get_rank_count(), std::forward<Args>(args)...);
+        ONEDAL_ASSERT(input_per_rank.size() == std::size_t(comm.get_rank_count()));
+
+        const auto local_result = dal::preview::train(comm, desc, input_per_rank[comm.get_rank()]);
+        return local_result;
+    }
+
+    template <typename Descriptor, typename... Args>
+    auto train_in_parallel_and_merge(const Descriptor& desc, Args&&... args) {
```

Comment:
In the future it would be good to think about the design that allow to remove code duplications in the methods like 'train_in_parallel_and_merge' and 'train_via_spmd_threads_and_merge'.

Reply:
I agree, but threading mode is completely different: single-proc-multi-thread against multi-proc-single-thread. Due to that train_via_spmd_threads_and_merge have to spawn multiple thread while train_in_parallel_and_merge is much simplier